content/en/platform/corda/5.0/developing-applications/api/ledger/utxo-ledger/vault-queries.md

```diff
@@ -0,0 +1,561 @@
+        )
+    }
+}
+```
+
+{{% /tab %}}
+{{% tab name="Java" %}}
+
+```java
+class DummyCustomQueryCollector implements VaultNamedQueryCollector<String, Integer> {
+    @NotNull
+    @Override
+    public Result<Integer> collect(@NotNull List<String> resultSet, @NotNull Map<String, Object> parameters) {
+        return new Result<>(
+                List.of(resultSet.size()),
+                true
+        );
+    }
+}
+```
+
+{{% /tab %}}
+{{< /tabs >}}
+
+{{< note >}}
+ The query `isDone` should only be set to true if the result set is complete.
+{{< /note >}}
+
+#### Registering our Complex Query
+
+ Register a complex query with a filter, a transformer, and a collector with the following example:
+
+{{< tabs name="tabs-7" >}}
+{{% tab name="Kotlin" %}}
+
+```kotlin
+class DummyCustomQueryFactory : VaultNamedQueryFactory {
+    override fun create(vaultNamedQueryBuilderFactory: VaultNamedQueryBuilderFactory) {
+        vaultNamedQueryBuilderFactory.create("DUMMY_CUSTOM_QUERY")
+            .whereJson(
+                "WHERE visible_states.custom_representation -> 'com.r3.corda.demo.contract.TestState' " +
+                        "->> 'testField' = :testField"
+            )
+            .filter(DummyCustomQueryFilter())
+            .map(DummyCustomQueryMapper())
+            .collect(DummyCustomQueryCollector())
+            .register()
+    }
+}
+```
+
+{{% /tab %}}
+{{% tab name="Java" %}}
+
+```java
+public class JsonQueryFactory implements VaultNamedQueryFactory {
+    @Override
+    public void create(@NotNull VaultNamedQueryBuilderFactory vaultNamedQueryBuilderFactory) {
+        vaultNamedQueryBuilderFactory.create("DUMMY_CUSTOM_QUERY")
+                .whereJson(
+                        "WHERE visible_states.custom_representation -> 'com.r3.corda.demo.contract.TestState' " +
+                                "->> 'testField' = :testField"
+                )
+                .filter(new DummyCustomQueryFilter())
+                .map(new DummyCustomQueryMapper())
+                .collect(new DummyCustomQueryCollector())
+                .register();
+    }
+}
+```
+
+{{% /tab %}}
+{{< /tabs >}}
+
+{{< note >}}
+The collector always needs to be the last one in the chain as all previous filtering and mapping functionality will be applied before collecting.
+{{< /note >}}
+
+### Executing a Vault-Named Query
+
+To execute a query use `UtxoLedgerService`. This can be injected to a flow via `@CordaInject`.
+To instantiate a query call the following:
+
+{{< tabs name="tabs-8" >}}
+{{% tab name="Kotlin" %}}
+
+```kotlin
+utxoLedgerService.query("DUMMY_CUSTOM_QUERY", Int::class.java)
+```
+
+{{% /tab %}}
+{{% tab name="Java" %}}
+
+```java
+utxoLedgerService.query("DUMMY_CUSTOM_QUERY", Integer.class)
+```
+
+{{% /tab %}}
+{{< /tabs >}}
+
+Provide the name of the query (in this case `DUMMY_CUSTOM_QUERY`) and the return type. Since the result set is collected into an integer in the complex query example use `Int` (or `Integer` in Java).
```

Comment:
```suggestion
Provide the name of the query (in this case `DUMMY_CUSTOM_QUERY`) and the return type. Since the result set is collected into an integer in the complex query example, use `Int` (or `Integer` in Java).
```